File: frontend/src/api/optionMetrics.ts
// Typed client + react-query hook for `GET /options/metrics` (SO-198).
//
// Stateless Black-Scholes greeks, implied vol, fair value, and break-even for a
// single call. The browser holds live spot (Pyth) and the order-book mid
// (DeepBook devInspect) and passes them in; the endpoint is pure compute.
//
// Authoritative response shape: rust-backend/services/api-service/README.md
// (`GET /options/metrics`). UNITS: spot/strike/mark must share one unit.

import { keepPreviousData, useQuery } from "@tanstack/react-query";

const API_BASE_URL: string =
  (import.meta.env.VITE_API_BASE_URL as string | undefined) ?? "http://127.0.0.1:9003";

export type OptionMetrics = {
  implied_vol: number | null;
  delta: number;
  gamma: number;
  vega: number; // per 1.00 vol — divide by 100 for per-1% display
  theta: number; // per calendar day
  rho: number; // per 1.00 rate
  break_even: number;
  fair_value: number | null;
};

export type MetricsParams = {
  spot: number;
  strike: number;
  t_years: number;
  mark: number; // order-book mid (quote) or avg cost (open position)
  r?: number;
};

export async function fetchOptionMetrics(p: MetricsParams): Promise<OptionMetrics> {
  const q = new URLSearchParams({
    spot: String(p.spot),
    strike: String(p.strike),
    t_years: String(p.t_years),
    mark: String(p.mark),
    ...(p.r != null ? { r: String(p.r) } : {}),
  });
  const res = await fetch(`${API_BASE_URL}/options/metrics?${q}`);
  if (!res.ok) throw new Error(`option metrics ${res.status}`);
  return res.json();
}

/** Years to expiry from an epoch-ms expiry. Floored at 0. */
export function yearsToExpiry(expiryMs: number, nowMs = Date.now()): number {
  return Math.max(0, (expiryMs - nowMs) / (365 * 24 * 60 * 60 * 1000));
}

/**
 * Live greeks/IV for a bucket. Re-fetches when spot or mid move (they're in the
 * queryKey) and on a ~1.5s poll so theta/τ stay fresh. Disabled until we have
 * spot + a positive mid.
 */
export function useOptionMetrics(args: {
  strike: number | null;
  expiryMs: number | null;
  spot: number | null;
  mid: number | null;
  r?: number;
  // Chain-table rows compute greeks for every strike; let them poll gently
  // (the selected strike keeps the default 1.5s cadence) — SO-225.
  refetchInterval?: number;
}) {
  const { strike, expiryMs, spot, mid, r, refetchInterval = 1500 } = args;
  const enabled =
    strike != null && expiryMs != null && spot != null && spot > 0 && mid != null && mid > 0;
  return useQuery({
    queryKey: ["option-metrics", strike, expiryMs, spot, mid, r ?? null],
    enabled,
    // Spot/mid live in the key, so every Pyth tick is a fresh query. Keep the
    // last result on screen while it refetches so greeks don't flash to "—"
    // between ticks (SO-225).
    placeholderData: keepPreviousData,
    refetchInterval,
    queryFn: () =>
      fetchOptionMetrics({
        spot: spot!,
        strike: strike!,
        t_years: yearsToExpiry(expiryMs!),
        mark: mid!,
        r,
      }),
  });
}

File: rust-backend/services/solana-api-service/src/handlers/option_metrics.rs
//! `GET /options/metrics` — Black-Scholes greeks, implied vol, fair value,
//! and break-even for a single call. STATELESS: every input is a query
//! param, so the service stays a pure compute endpoint (no Pyth / venue
//! reads). The frontend holds live spot and the observed option mark and
//! passes them in. Ported verbatim from the Sui twin — the `pricing` crate
//! is chain-agnostic.
//!
//! UNITS: `spot`, `strike`, and `mark` MUST share one unit (e.g. settlement
//! per option, whole units). `t_years` is calendar time to expiry in years.
//! `r` is the annualized continuous risk-free rate; omit for the default.

use axum::{extract::Query, Json};
use serde::{Deserialize, Serialize};

use pricing::{break_even, call_greeks, call_price_per_unit, implied_vol, CallInputs};

/// Default risk-free rate when `r` is omitted. 0 keeps greeks comparable
/// to a retail screen that ignores rates.
const DEFAULT_RATE: f64 = 0.0;

#[derive(Deserialize)]
pub struct MetricsQuery {
    pub spot: f64,
    pub strike: f64,
    pub t_years: f64,
    /// Observed option price (order-book mid, or avg cost for break-even),
    /// same unit as spot/strike.
    pub mark: f64,
    pub r: Option<f64>,
}

#[derive(Serialize)]
pub struct MetricsResponse {
    /// Annualized vol, e.g. 0.3492 for 34.92%. `null` if no solution
    /// (mark below intrinsic, ≥ spot, or expiry ≤ 0).
    pub implied_vol: Option<f64>,
    pub delta: f64,
    pub gamma: f64,
    pub vega: f64,
    /// Per-calendar-day theta.
    pub theta: f64,
    pub rho: f64,
    pub break_even: f64,
    /// BS price re-priced at the solved IV — a sanity echo of `mark`.
    /// `null` when there is no IV.
    pub fair_value: Option<f64>,
}

pub async fn option_metrics(Query(q): Query<MetricsQuery>) -> Json<MetricsResponse> {
    let r = q.r.unwrap_or(DEFAULT_RATE);
    let iv = implied_vol(q.mark, q.spot, q.strike, q.t_years, r);

    // Greeks at the solved IV. With no IV (degenerate expiry / out-of-band
    // mark) fall back to sigma=0 so the row still renders (greeks collapse
    // to the deterministic case).
    let sigma = iv.unwrap_or(0.0);
    let inputs = CallInputs { spot: q.spot, strike: q.strike, t_years: q.t_years, r, sigma };
    let g = call_greeks(inputs);
    let fair_value = iv.map(|s| call_price_per_unit(CallInputs { sigma: s, ..inputs }));

    Json(MetricsResponse {
        implied_vol: iv,
        delta: g.delta,
        gamma: g.gamma,
        vega: g.vega,
        theta: g.theta,
        rho: g.rho,
        break_even: break_even(q.strike, q.mark),
        fair_value,
    })
}
